server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileManager.java

```diff
@@ -571,4 +583,98 @@ public MetadataUpdateCount getUpdateCount() {
+   * still out of sync, then recovery is performed dependent on the action: 1) log: nothing else is
+   * done 2) logsync: memory is synched with the metadata IFF the operation log agrees with one or
+   * the other 3) metasync: metadata files are reloaded into memory IFF the operation log agrees
+   * with metadata 4) memsync: metadata is reset with the memory files IFF the operation log agrees
+   * with memory
+   *
+   * @param context The server context
+   */
+  public void handleMetadataDiff(ServerContext context) {
+    LogSync action =
+        LogSync.valueOf(tablet.getTableConfiguration().get(Property.TABLE_OPERATION_LOG_RECOVERY));
+    synchronized (tablet) {
+      // always log the operation log regardless of the requested action
+      log.error("Operation log: " + tabletLog.dumpLog());
+
+      // rescan for the tablet metadata
+      var tabletMeta =
+          context.getAmple().readTablet(tablet.getExtent(), TabletMetadata.ColumnType.FILES);
+      if (tabletMeta == null) {
+        String msg = "Tablet " + tablet.getExtent() + " not found in metadata";
+        log.error(msg);
+      } else {
+        Set<StoredTabletFile> metadata = new HashSet<>(tabletMeta.getFiles());
+        Set<StoredTabletFile> memory = new HashSet<>(datafileSizes.keySet());
+        Set<StoredTabletFile> expected = tabletLog.getExpectedFiles();
+
+        // verify we are still out of sync
+        if (metadata.equals(memory)) {
+          log.debug("Metadata and in-memory file list are back in-sync: " + metadata);
+          if (!expected.equals(memory)) {
+            log.error(
+                "Resetting operation log " + expected + " with metadata and memory " + memory);
+            tabletLog.reset(memory);
+          }
+        } else if (!action.equals(LogSync.log)) {
+          if (action.equals(LogSync.logsync)) {
+            if (expected.equals(memory)) {
+              action = LogSync.memsync;
+            } else if (expected.equals(metadata)) {
+              action = LogSync.metasync;
+            } else {
+              log.error("Not synching files because the operation log " + expected
+                  + " does not agree with metadata " + metadata + " nor memory " + memory);
+            }
+          }
+
+          if (action.equals(LogSync.metasync)) {
+            if (!expected.equals(metadata)) {
+              log.error("Not synching memory because the operation log " + expected
+                  + " does not agree with metadata " + metadata);
+            } else {
+              resetMemoryWithMetadata(tabletMeta.getFilesMap());
+            }
+          } else if (action.equals(LogSync.memsync)) {
+            if (!expected.equals(memory)) {
+              log.error("Not synching metadata because the operation log " + expected
+                  + " does not agree with memory " + memory);
+            } else {
+              resetMetadataWithMemory(metadata, datafileSizes);
+            }
+          }
```

Comment:
In the `elasticity` branch there is API for refreshing the tablet when the metadata changes. This was added because the tablet management is being performed in the Manager. Looking at that code, I think there may be more to do here. Take a look at [TabletClientHandler.refreshTablets](https://github.com/apache/accumulo/blob/elasticity/server/tserver/src/main/java/org/apache/accumulo/tserver/TabletClientHandler.java#L1145) and [Tablet.refreshMetadata](https://github.com/apache/accumulo/blob/elasticity/server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java#L1453)

Reply:
note that this is written for 2.x, not 3.x

Reply:
Right, but I still think there are more actions that need to be taken, even in 2.1, based on the similar functionality.